File: internal/maintenance/cat_caches.go
package maintenance

import (
	"context"
	"os"
	"path/filepath"
)

func init() { registerMatcher("caches", 0, scanCaches) }

// cacheSurfaces is the KNOWN-SURFACE ALLOWLIST — never a *cache* glob. Each
// entry justifies its clearability with who rebuilds it; a file with no note
// gets no clear button (it isn't listed here).
var cacheSurfaces = []struct{ rel, regeneratedBy string }{
	{"cache/changelog.md", "CLI update check"},
	{"stats-cache.json", "usage tracking"},
	{"mcp-needs-auth-cache.json", "next MCP probe"},
}

// scanCaches surfaces the small regenerable caches (plain-delete). paste-cache
// blobs may hold pasted secrets — flagged sensitive and, being regenerable,
// plain-deleted (no trash copy that would extend retention).
func scanCaches(ctx context.Context, spec CategorySpec) ([]Candidate, error) {
	out := []Candidate{}
	for _, s := range cacheSurfaces {
		p := filepath.Join(spec.Root, s.rel)
		info, err := os.Lstat(p)
		if err != nil || info.IsDir() || info.Mode()&os.ModeSymlink != 0 {
			continue
		}
		out = append(out, Candidate{
			Path: p, Bytes: info.Size(), Files: 1, ModTime: info.ModTime(),
			Reason: "cache — rebuilt on demand",
			Meta:   map[string]string{"regeneratedBy": s.regeneratedBy},
		})
	}

	pasteDir := filepath.Join(spec.Root, "paste-cache")
	entries, ok, err := openDirNoSymlink(pasteDir)
	if err != nil {
		return nil, err
	}
	if ok {
		for _, e := range entries {
			if err := ctx.Err(); err != nil {
				return nil, err
			}
			if e.IsDir() {
				continue
			}
			info, infoErr := e.Info()
			if infoErr != nil || info.Mode()&os.ModeSymlink != 0 {
				continue
			}
			out = append(out, Candidate{
				Path: filepath.Join(pasteDir, e.Name()), Bytes: info.Size(), Files: 1,
				ModTime: info.ModTime(), Reason: "pasted content — may contain sensitive text",
				Group: "paste-cache",
				Meta:  map[string]string{"regeneratedBy": "next paste", "sensitive": "true"},
			})
		}
	}
	return out, nil
}

File: src-tauri/src/maintenance/cat_caches.rs
//! Ported from `internal/maintenance/cat_caches.go` (W13). Surfaces the small
//! regenerable caches (plain-delete) via a KNOWN-SURFACE ALLOWLIST, plus
//! paste-cache blobs (flagged sensitive — may hold pasted secrets).

use std::collections::BTreeMap;
use std::fs;
use std::path::Path;

use super::category::{mtime_utc, open_dir_no_symlink};
use super::types::{Candidate, CategorySpec};

/// The KNOWN-SURFACE ALLOWLIST — never a `*cache*` glob. `(rel, regenerated_by)`.
/// Mirrors Go `cacheSurfaces`.
const CACHE_SURFACES: &[(&str, &str)] = &[
    ("cache/changelog.md", "CLI update check"),
    ("stats-cache.json", "usage tracking"),
    ("mcp-needs-auth-cache.json", "next MCP probe"),
];

pub fn scan_caches(spec: &CategorySpec) -> Result<Vec<Candidate>, String> {
    let root = Path::new(&spec.root);
    let mut out: Vec<Candidate> = Vec::new();

    for (rel, regenerated_by) in CACHE_SURFACES {
        let p = root.join(rel);
        let info = match fs::symlink_metadata(&p) {
            Ok(i) => i,
            Err(_) => continue,
        };
        if info.is_dir() || info.file_type().is_symlink() {
            continue;
        }
        let mut meta = BTreeMap::new();
        meta.insert("regeneratedBy".to_string(), (*regenerated_by).to_string());
        out.push(Candidate {
            path: p.to_string_lossy().into_owned(),
            bytes: info.len() as i64,
            files: 1,
            mod_time: mtime_utc(&info),
            reason: "cache — rebuilt on demand".to_string(),
            group: String::new(),
            meta,
        });
    }

    let paste_dir = root.join("paste-cache");
    let (entries, ok) = open_dir_no_symlink(&paste_dir)?;
    if ok {
        for e in &entries {
            if e.file_type().map(|t| t.is_dir()).unwrap_or(false) {
                continue;
            }
            let Ok(info) = e.metadata() else {
                continue;
            };
            if info.file_type().is_symlink() {
                continue;
            }
            let name = e.file_name().to_string_lossy().into_owned();
            let mut meta = BTreeMap::new();
            meta.insert("regeneratedBy".to_string(), "next paste".to_string());
            meta.insert("sensitive".to_string(), "true".to_string());
            out.push(Candidate {
                path: paste_dir.join(&name).to_string_lossy().into_owned(),
                bytes: info.len() as i64,
                files: 1,
                mod_time: mtime_utc(&info),
                reason: "pasted content — may contain sensitive text".to_string(),
                group: "paste-cache".to_string(),
                meta,
            });
        }
    }
    Ok(out)
}

#[cfg(test)]
#[path = "cat_logs_caches_tests.rs"]
mod cat_logs_caches_tests;
